File: cocsim/dto_game_renderer.py
from typing import TypedDict

from . import game
from .shapes import Shape


class Frame(TypedDict):
    timeElapsed: float
    progressInfo: str
    totalBaseSize: int
    grid: list[Shape] | None  # None if not changed
    collision: list[Shape] | None  # None if not changed
    entities: list[Shape]


class DTOGameRenderer:
    result: list[Frame]
    ticks_per_draw: int

    _ticks_since_last_draw: int | None

    def __init__(self, ticks_per_draw: int):
        self.result = []
        self.ticks_per_draw = ticks_per_draw
        self._ticks_since_last_draw = None

    def draw(self, game: "game.Game"):
        if (
            self._ticks_since_last_draw is None
            or self._ticks_since_last_draw == self.ticks_per_draw
        ):
            self._ticks_since_last_draw = 0

            self._draw(game)

        self._ticks_since_last_draw += 1

    def finish(self, game: "game.Game"):
        if self.result[-1]["timeElapsed"] != game.time_elapsed:
            self._draw(game)

    def _draw(self, game: "game.Game"):
        frame: Frame = {
            "timeElapsed": game.time_elapsed,
            "progressInfo": game.progress_info(),
            "entities": game.draw_entities(),
            "totalBaseSize": game.total_size,
        }

        if len(self.result) == 0:
            frame["grid"] = game.draw_grid()
        else:
            frame["grid"] = None

        if game.need_redraw_collision():
            frame["collision"] = game.draw_collision()
        else:
            frame["collision"] = None

        self.result.append(frame)

File: backend/api_base/src/dto_game_renderer.rs
use cocsim::{
    Game,
    Shape,
};
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Frame {
    pub time_elapsed: f32,
    pub progress_info: String,
    pub total_base_size: usize,
    /// None if not changed
    pub grid: Option<Vec<Shape>>,
    /// None if not changed
    pub collision: Option<Vec<Shape>>,
    pub entities: Vec<Shape>,
}

pub struct DtoGameRenderer {
    result: Vec<Frame>,
    ticks_per_draw: usize,
    ticks_since_last_draw: Option<usize>,
}

impl DtoGameRenderer {
    pub fn new(ticks_per_draw: usize) -> Self {
        Self {
            result: Vec::new(),
            ticks_per_draw,
            ticks_since_last_draw: None,
        }
    }

    pub fn draw(&mut self, game: &mut Game) {
        if self.ticks_since_last_draw.is_none()
            || self.ticks_since_last_draw == Some(self.ticks_per_draw)
        {
            self.ticks_since_last_draw = Some(0);

            self.draw_internal(game);
        }

        *self.ticks_since_last_draw.as_mut().unwrap() += 1;
    }

    pub fn finish(mut self, game: &mut Game) -> Vec<Frame> {
        if self.result.last().unwrap().time_elapsed != game.time_elapsed() {
            self.draw_internal(game);
        }

        self.result
    }

    fn draw_internal(&mut self, game: &mut Game) {
        let grid = if self.result.is_empty() {
            Some(game.draw_grid())
        } else {
            None
        };
        let collision = if game.need_redraw_collision() {
            Some(game.draw_collision())
        } else {
            None
        };

        self.result.push(Frame {
            time_elapsed: game.time_elapsed(),
            progress_info: game.progress_info(),
            total_base_size: game.map_size().total_size() as usize,
            grid,
            collision,
            entities: game.draw_entities(),
        });
    }
}
